File: frontend/types/grievance.ts
export type Priority = 'critical' | 'high' | 'medium' | 'low';

export type Status = 'pending' | 'classified' | 'in_progress' | 'resolved' | 'closed';

export interface Grievance {
  id: number;
  title: string;
  description: string;
  category: string;
  subcategory?: string;
  ward_id: number;
  ward_name: string;
  ward?: string;
  lat: number;
  lng: number;
  status: Status;
  priority: Priority;
  score: number;
  sentiment: string;
  source: string;
  created_at: string;
  updated_at: string;
}

export interface Ward {
  id: number;
  name: string;
}

File: backend/app/models/grievance.py
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class Grievance(Base):
    __tablename__ = "grievances"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    category: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    subcategory: Mapped[str | None] = mapped_column(String(100), nullable=True)
    ward_id: Mapped[int] = mapped_column(ForeignKey("wards.id"), nullable=False, index=True)
    ward_name: Mapped[str] = mapped_column(String(100), nullable=False)
    lat: Mapped[float] = mapped_column(Float, nullable=False)
    lng: Mapped[float] = mapped_column(Float, nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending", index=True)
    priority: Mapped[str] = mapped_column(String(20), nullable=False, default="low", index=True)
    score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    sentiment: Mapped[str] = mapped_column(String(20), nullable=False, default="neutral")
    source: Mapped[str] = mapped_column(String(20), nullable=False, default="csv")
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, server_default=func.now())
    updated_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True, onupdate=func.now())
